### ISSUE TITLE
the model trained by same datasets created by same file seems not be the same.

### ISSUE THREAD

**Issue by wbo4958** [dmlc/xgboost#6894]
1. data file: a.csv
2. parameters: PARAM
2. create dmatrix1 from **a.csv** by XGDeviceQuantileDMatrixCreateFromCallback
3. create dmatrix2 from **a.csv** by XGDMatrixCreateFromArrayInterfaceColumns
4. model1 = train(dmatrix1,  PARAM)
5. model2 = train(dmatrix2. PARAM)
6. predict1 = predict(dmatrix2)
7. predict2 = predict(dmatrix2)

but predict1 != predict2, we expect predict1==predict2.

The issue is a regression caused by https://github.com/dmlc/xgboost/pull/6860

@trivialfis Could you have a chance to look at it?

**Comment by trivialfis** [comment]
I couldn't reproduce:

``` python
import cupy
import numpy as np
import xgboost
from sklearn.datasets import make_classification
import cudf


def classification():
    X, y = make_classification(
        n_samples=10000, n_features=10, n_classes=8, n_informative=8
    )
    X = cudf.DataFrame(X)
    y = cudf.Series(y)

    Xy = xgboost.DMatrix(X, y)
    dXy = xgboost.DeviceQuantileDMatrix(X, y)

    m_Xy = xgboost.train(
        {"tree_method": "gpu_hist", "num_class": 8}, Xy, num_boost_round=100
    )
    m_dXy = xgboost.train(
        {"tree_method": "gpu_hist", "num_class": 8}, dXy, num_boost_round=100
    )

    p = m_Xy.predict(Xy)
    dp = m_dXy.predict(dXy)
    np.testing.assert_allclose(p, dp)


def regression():
    rng = cupy.random.RandomState(100)
    X = rng.randn(10000, 10)
    y = rng.randn(10000)
    X = cudf.DataFrame(X)
    y = cudf.Series(y)

    Xy = xgboost.DMatrix(X, y)
    dXy = xgboost.DeviceQuantileDMatrix(X, y)

    m_Xy = xgboost.train({"tree_method": "gpu_hist"}, Xy, num_boost_round=100)
    m_dXy = xgboost.train({"tree_method": "gpu_hist"}, dXy, num_boost_round=100)

    p = m_Xy.predict(Xy)
    dp = m_dXy.predict(dXy)
    np.testing.assert_allclose(p, dp)


if __name__ == "__main__":
    classification()
    regression()
```

**Comment by wbo4958** [comment]
The trained models are completely the same. the only difference is one predict used cached predication, while the other one was calculating runtime by GPU. So there may be some float accuracy.

Close this issue. thx @trivialfis
